### ISSUE TITLE
Error while saving bot state to azure blob storage

### ISSUE THREAD

**Issue by harishk3491** [microsoft/botbuilder-dotnet#6590]
am trying to save bot state to azure blob storage using nuget package Microsoft.Bot.Builder.Azure.Blobs version 4.19.2 for .NET. What i am doing is in Startup.cs file is

var storage = new BlobsStorage( "connectionstring", "containername" );

var userState = new UserState(storage); services.AddSingleton(userState);

var conversationState = new ConversationState(storage); services.AddSingleton(conversationState);

But i am getting following error

System.InvalidOperationException: Unable to find the following types in the 'AllowedTypes' collection.

Please provide the 'AllowedTypesSerializationBinder' in the custom 'JsonSerializerSettings' instance, with the list of types to allow.

Example: new JsonSerializerSettings { SerializationBinder = new AllowedTypesSerializationBinder( new List { typeof(PromptOptions), typeof(DialogInstance), typeof(DialogState), }), }

When i check container blob files are generated anyways

**Comment by anishprasad01** [comment]
Hi @harishk3491,

Based on the code you have provided, it's possible that the setup code you have is a bit out of date. Try using the code from the [documentation](https://learn.microsoft.com/en-us/azure/bot-service/bot-builder-howto-v4-storage?view=azure-bot-service-4.0&tabs=csharp#add-blob-storage-configuration-information) and see if that fixes the error.

**Comment by harishk3491** [comment]
I changed code based on above link 
 services.AddSingleton<IStorage>(
            new BlobsStorage(
                Configuration.GetValue<string>("BlobConnectionString"),
                Configuration.GetValue<string>("BlobContainerName"),
                new JsonSerializer()
                {
                    SerializationBinder = new AllowedTypesSerializationBinder(
            new List<Type>
            {
                typeof(PromptOptions),
                typeof(DialogInstance),
                typeof(DialogState),
            })
                }
                ));

            services.AddSingleton<UserState>();
            services.AddSingleton<ConversationState>();




now its giving me another error System.NullReferenceException' in Microsoft.Bot.Builder.Dialogs.dll
Object reference not set to an instance of an object. 

![Screenshot 2023-02-18 012104](https://user-images.githubusercontent.com/11428333/219777229-ceb26f79-7ba1-4ad8-ab12-92624ac94a24.png)

**Comment by anishprasad01** [comment]
I'm sorry, I may have caused a bit of confusion. You do still need to pass in the state objects.

```csharp
// Create the User state passing in the storage layer.
var userState = new UserState(storage);
services.AddSingleton(userState);

// Create the Conversation state passing in the storage layer.
var conversationState = new ConversationState(storage);
services.AddSingleton(conversationState);
```

**Comment by harishk3491** [comment]
I am getting this issue when using Microsoft.Bot.Builder.Azure.Blobs version **4.19.2** if i revert back to **4.18.1** its working properly. 
![23](https://user-images.githubusercontent.com/11428333/219810513-f6ad8010-8ce2-4ad4-934a-abf8c64005a5.png)

**Comment by anishprasad01** [comment]
I'll investigate further.

**Comment by harishk3491** [comment]
I updated all bot packages to 4.19.2 
![100](https://user-images.githubusercontent.com/11428333/220189292-861480ce-898a-4ae1-83b0-0084f8d96d8d.png)
![101](https://user-images.githubusercontent.com/11428333/220189309-08ce126c-7d77-4e01-ab9e-fd8f90234370.png)
![102](https://user-images.githubusercontent.com/11428333/220189312-10ddcb54-d7a2-43a2-abe7-78e6090e7597.png)

**Comment by anishprasad01** [comment]
What other information can you provide about your setup? Do you experience this error with the sample on 4.19.2?

**Comment by harishk3491** [comment]
This is the code base i am working [https://github.com/harishk3491/DemoBot](url) 
Both blob storage and cosmos will work if i change the version to **4.18.1** but get above error in **4.19.2**

**Comment by ErwanL08** [comment]
I have the same issue,  the bot send back his information and when the state is saved it throw this exception (work fine with 4.18.1)
![image](https://user-images.githubusercontent.com/16559628/220315174-87824ffd-add1-4192-87ae-6b65c89f111e.png)
in startup 
![image](https://user-images.githubusercontent.com/16559628/220315542-b104ef2a-9e16-44fc-9793-c13e9358393f.png)

**Comment by anishprasad01** [comment]
@harishk3491 @ErwanL08,

Since the error message is asking us for a custom binder, let's go ahead and create one. 

Please try the following:

1. Create a new `JsonSerializer` variable in `Startup.cs`, such as in the following example:
```csharp
var serializer = new JsonSerializer 
{
    TypeNameHandling = TypeNameHandling.Objects,
    MaxDepth = null,
    SerializationBinder = new AllowedTypesSerializationBinder(
            new List<Type>
            {
                typeof(string),
                typeof(Dictionary<string, object>),
                typeof(Activity)
            }),
};
```

**Comment by anishprasad01** [comment]
@harishk3491 @ErwanL08,

Since the error message is asking us for a custom binder, let's go ahead and create one. 

Please try the following:

1. Create a new `JsonSerializer` variable in `Startup.cs`, such as in the following example:
```csharp
var customSerializer = new JsonSerializer 
{
    TypeNameHandling = TypeNameHandling.Objects,
    MaxDepth = null,
    SerializationBinder = new AllowedTypesSerializationBinder(
            new List<Type>
            {
                typeof(string),
                typeof(Dictionary<string, object>),
                typeof(Activity)
            }),
};
```
2. Pass the new serializer to the BlobsStorage class, like this:
```csharp
var storage = new BlobsStorage("<BlobConnectionString>", "<BlobContainerName>", customSerializer);
```
3. Test the bot and observe the response or error message.
4. Add to the `List<Type>` in the customSerializer any types you have added or created in your project that you wish to serialize and send to Blob storage.

Let me know the results.

**Comment by harishk3491** [comment]
@anishprasad01  Above solution works. I included following types for my particular project.

` var customSerializer = new JsonSerializer
            {
                TypeNameHandling = TypeNameHandling.Objects,
                MaxDepth = null,
                SerializationBinder = new AllowedTypesSerializationBinder(
            new List<Type>
            {
                typeof(PromptOptions),
                typeof(DialogInstance),
                typeof(DialogState),
                typeof(Dictionary<string, object>),
                typeof(Choice),
            }),
            };

            services.AddSingleton<IStorage>(
            new BlobsStorage(
                Configuration.GetValue<string>("BlobConnectionString"),
                Configuration.GetValue<string>("BlobContainerName"),customSerializer
                ));


            services.AddSingleton<UserState>();
            services.AddSingleton<ConversationState>();`

**Comment by anishprasad01** [comment]
@nimbusparis

It would be helpful for my investigation if you could tell me which samples you tried that produced this error.

**Comment by anishprasad01** [comment]
@ceciliaavila,

This issue seems to stem from #6549 and is related to the attempted fixes in #6585. It appears that those fixes did not prevent the error requiring the addition of a custom serializer. Fixing this will rpobably require investigation into whether the original CodeQL errors should be fixed in a different manner.

**Comment by tracyboehrer** [comment]
@nimbusparis We published a test release with the related commits reverted.  Would you be willing to test this release?

You will need to add a Package Source for: 
https://pkgs.dev.azure.com/ConversationalAI/BotFramework/_packaging/SDK/nuget/v3/index.json

The version number is:  4.19.3-rc0

**Comment by tracyboehrer** [comment]
Patched in 4.19.3
